Mathlib/Data/Fin/Basic.lean

```diff
@@ -1865,54 +1980,107 @@ def predAbove (p : Fin n) (i : Fin (n + 1)) : Fin n :=
+    p.predAbove (succ i) = (i.succ).castPred hi := by
+  rw [predAbove_of_lt_succ _ _ (succ_lt_succ_iff.mpr h)]
+theorem predAbove_succ_of_le (p i : Fin n) (h : p ≤ i) :
+    p.predAbove (succ i) = i := by
+  rw [predAbove_of_succ_le _ _ (succ_le_succ_iff.mpr h), pred_succ]
+@[simp]
+theorem predAbove_succ_self (p : Fin n) : p.predAbove (succ p) = p :=
+  predAbove_succ_of_le _ _ le_rfl
+
+theorem predAbove_castSucc_of_lt (p i : Fin n) (h : p < i) (hi := castSucc_ne_zero_of_lt h) :
+    p.predAbove (castSucc i) = i.castSucc.pred hi := by
+  rw [predAbove_of_castSucc_lt _ _ (castSucc_lt_castSucc_iff.mpr h)]
+theorem predAbove_castSucc_of_le (p i : Fin n) (h : i ≤ p) :
+    p.predAbove (castSucc i) = i := by
+  rw [predAbove_of_le_castSucc _ _ (castSucc_le_castSucc_iff.mpr h), castPred_castSucc]
+@[simp]
+theorem predAbove_castSucc_self (p : Fin n) : p.predAbove (castSucc p) = p :=
+  predAbove_castSucc_of_le _ _ le_rfl
+
+theorem predAbove_pred_of_lt (p i : Fin (n + 1)) (h : i < p) (hp := ((zero_le i).trans_lt h).ne')
+    (hi := ((le_last p).trans_lt' h).ne) : (pred p hp).predAbove i = castPred i hi := by
+  rw [predAbove_of_lt_succ _ _ (succ_pred _ _ ▸ h)]
+theorem predAbove_pred_of_le (p i : Fin (n + 1)) (h : p ≤ i) (hp : p ≠ 0)
+    (hi := (h.trans_lt' (pos_of_ne_zero hp)).ne') : (pred p hp).predAbove i = pred i hi := by
+  rw [predAbove_of_succ_le _ _ (succ_pred _ _ ▸ h)]
+theorem predAbove_pred_self (p : Fin (n + 1)) (hp : p ≠ 0) :
+    (pred p hp).predAbove p = pred p hp := predAbove_pred_of_le _ _ le_rfl hp
+
+theorem predAbove_castPred_of_lt (p i : Fin (n + 1)) (h : p < i)
+    (hp := ((le_last i).trans_lt' h).ne) (hi := ((zero_le p).trans_lt h).ne') :
+    (castPred p hp).predAbove i = pred i hi := by
+  rw [predAbove_of_castSucc_lt _ _ (castSucc_castPred _ _ ▸ h)]
+theorem predAbove_castPred_of_le (p i : Fin (n + 1)) (h : i ≤ p) (hp : p ≠ last n)
+    (hi := (h.trans_lt (lt_top_iff_ne_top.mpr hp)).ne) :
+    (castPred p hp).predAbove i = castPred i hi := by
+  rw [predAbove_of_le_castSucc _ _ (castSucc_castPred _ _ ▸ h)]
+theorem predAbove_castPred_self (p : Fin (n + 1)) (hp : p ≠ last n) :
+    (castPred p hp).predAbove p = castPred p hp := predAbove_castPred_of_le _ _ le_rfl hp
+
+theorem predAbove_rev_left (p : Fin n) (i : Fin (n + 1)) :
+    p.rev.predAbove i = (p.predAbove i.rev).rev := by
+  obtain h | h := (rev i).succ_le_or_le_castSucc p
+  · rw [predAbove_of_succ_le _ _ h, rev_pred,
+      predAbove_of_le_castSucc _ _ (rev_succ _ ▸ (le_rev_iff.mpr h)), castPred_inj, rev_rev]
+  · rw [predAbove_of_le_castSucc _ _ h, rev_castPred,
+      predAbove_of_succ_le _ _ (rev_castSucc _ ▸ (rev_le_iff.mpr h)), pred_inj, rev_rev]
+theorem predAbove_rev_right (p : Fin n) (i : Fin (n + 1)) :
+    p.predAbove i.rev = (p.rev.predAbove i).rev := by
+  rw [predAbove_rev_left, rev_rev]
+
+@[simp]
+theorem predAbove_right_zero [NeZero n] {i : Fin n} : predAbove (i : Fin n) 0 = 0 := by
```

Comment:
I am a bit worried that all these `simp` lemmas will fire many TC synths for `NeZero n`. I will call a benchmark on this PR to see if it impacts the library.

Reply:
This is fair. I would be happy to have the `n + 1` versions as simp lemmas if that would be preferred? I had them in an earlier iteration of this work but I understood that NeZero n was preferred if possible.

Reply:
In this case, thinking about it the NeZero is only needed to state the lemma (because it is vacuously true for n = 0 because of i). So maybe I can change the statement somewhat? Essentially what I want is the zero on the right, and it's getting access to that notation that's the kicker.

Reply:
It does seem to impact it. Assuming that's what is causing the issue.

Reply:
It's in the approximate topic area, so it might be that whatever that file does it does so in a way that's very sensitive to this change?

Reply:
> Hmm, weird that only that one file would see such a negative impact. I wonder if it's a fluke? I'll bump master and run another bench.

Well, we've [seen](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/noisy.20.60FilteredColimitCommutesFiniteLimit.60) this [before](https://github.com/leanprover-community/mathlib4/pull/9655#issuecomment-1887793644).